### PR TITLE
Cast boolean values

### PR DESCRIPTION
**Pull request by i7an**
Boolean methods return raw values. AWS Parameters Store does not support boolean type.

```ruby
Global.foo.enabled? # => "false"

if Global.foo.enabled? # "false" is truthy
  # ...
end
```

In this PR I attempted to address this problem:

```ruby
Global.foo.enabled # => "false"
Global.foo.enabled.class # => String
Global.foo.enabled? # => false
Global.foo.enabled?.class # => FalseClass
```

### CODE REVIEW

File: Rakefile [hunk 1]
```diff
@@ -1,12 +1,7 @@
 # frozen_string_literal: true
 
-require 'rubygems'
-require 'bundler'
-
-Bundler.require
```
Comment (i7an) [inline Rakefile:6]: https://bundler.io/guides/bundler_setup.html

> For such a small Gemfile, we’d advise you to skip Bundler.require and just require the gems by hand

File: Rakefile [hunk 2]
```diff
@@ -1,12 +1,7 @@
 # frozen_string_literal: true
 
-require 'rubygems'
-require 'bundler'
```
Comment (i7an) [inline Rakefile:4]: Did not find why these were needed.

File: lib/global/backend/aws_parameter_store.rb [hunk 3]
```diff
@@ -95,7 +95,7 @@ def load_parameters_from_ssm(next_token = nil)
       def build_configuration_from_parameters(parameters)
         configuration = {}
         parameters.each do |parameter|
-          parameter_parts = parameter.name[@prefix.length..-1].split(PATH_SEPARATOR).map(&:to_sym)
+          parameter_parts = parameter.name[@prefix.length..].split(PATH_SEPARATOR).map(&:to_sym)
```
Comment (i7an) [inline lib/global/backend/aws_parameter_store.rb:98]: rubocop autofix. applies to other as well.

File: lib/global/configuration.rb [hunk 4]
```diff
@@ -13,6 +13,21 @@ class Configuration
...
+      false, 0,
+      '0', :'0',
+      'f', :f,
+      'F', :F,
+      'false', :false,
+      'FALSE', :FALSE,
+      'off', :off,
+      'OFF', :OFF
+    ].to_set.freeze
+    private_constant :FALSE_VALUES
```
Comment (i7an) [inline lib/global/configuration.rb:28]: `FALSE_VALUES` is implementation details and should not be exposed.

File: lib/global/configuration.rb [hunk 5]
```diff
@@ -49,9 +64,10 @@ def respond_to_missing?(method_name, include_private = false)
     end
 
     def method_missing(method, *args, &block)
-      method = normalize_key_by_method(method)
-      if key?(method)
-        get_configuration_value(method)
+      normalized_method = normalize_key_by_method(method)
+      if key?(normalized_method)
+        value = get_configuration_value(normalized_method)
+        boolean_method?(method) ? cast_boolean(value) : value
```
Comment (i7an) [inline lib/global/configuration.rb:70]: this is what this PR is all about.

File: spec/global/backend/aws_parameter_store_spec.rb [hunk 6]
```diff
@@ -1,18 +1,18 @@
 # frozen_string_literal: true
 
-require 'spec_helper'
 require 'aws-sdk-ssm'
 require 'global/backend/aws_parameter_store'
 
 RSpec.describe Global::Backend::AwsParameterStore do
+  subject(:parameter_store) do
+    described_class.new(prefix: '/testapp/', client: client)
+  end
```
Comment (i7an) [inline spec/global/backend/aws_parameter_store_spec.rb:9]: fixed rubocop abuses. applies to other spec files.

File: spec/spec_helper.rb [hunk 7]
```diff
@@ -1,9 +1,5 @@
 # frozen_string_literal: true
 
-$LOAD_PATH.unshift(File.join(File.dirname(__FILE__), '..', 'lib'))
-$LOAD_PATH.unshift(File.dirname(__FILE__))
```
Comment (i7an) [inline spec/spec_helper.rb:4]: `spec` and `lib` directories are in the load path by default

File: spec/spec_helper.rb [hunk 8]
```diff
@@ -1,9 +1,5 @@
...
-require 'rspec'
```
Comment (i7an) [inline spec/spec_helper.rb:5]: `rspec/core` is loaded by default

File: spec/spec_helper.rb [hunk 9]
```diff
@@ -1,9 +1,5 @@
...
-require 'global'
```
Comment (i7an) [inline spec/spec_helper.rb:6]: `simplecov` should be required before any code is loaded
https://github.com/simplecov-ruby/simplecov?tab=readme-ov-file#troubleshooting

**Review by leonid-shevtsov — APPROVED** [review]
I suppose the ruby version range is acceptable, seeing as 3.0 is already EOL

File: .github/workflows/tests.yml [hunk 10]
```diff
@@ -10,7 +10,11 @@ jobs:
     strategy:
       matrix:
         os: [ubuntu, macos]
-        ruby-version: ['3.1', '3.0', '2.7', '2.6', '2.5']
+        ruby-version:
+          - 3.3
+          - 3.2
+          - 3.1
+          - 3.0
```
Comment (i7an) [inline .github/workflows/tests.yml:17]: image

File: spec/global/configuration_spec.rb [hunk 11]
```diff
@@ -1,163 +1,135 @@
...
 RSpec.describe Global::Configuration do
-  let(:hash) { { 'key' => 'value', 'boolean_key' => true, 'nested' => { 'key' => 'value' }} }
-  let(:configuration) { described_class.new hash }
+  subject(:configuration) { described_class.new hash }
+
+  let(:hash) do
+    {
+      'key' => 'value',
+      'boolean_key' => true,
+      'string_boolean_key' => '1',
```
Comment (le0pard) [inline spec/global/configuration_spec.rb:10]: test with empty string and false value

File: spec/global_merge_backends_spec.rb [hunk 12]
```diff
@@ -0,0 +1,25 @@
+# frozen_string_literal: true
+
+RSpec.describe Global do
+  subject(:global) { described_class }
+
+  describe 'merging backends' do
+    # rubocop:disable RSpec/VerifiedDoubles
```
Comment (le0pard) [inline spec/global_merge_backends_spec.rb:7]: this can be fixed and written correctly

**Review by le0pard — COMMENTED** [review]
need some fixes

File: global.gemspec [hunk 13]
```diff
@@ -1,31 +1,24 @@
...
   s.authors = ['Railsware LLC']
   s.email = 'contact@railsware.com'
-
   s.description = 'Simple way to load your configs from yaml/aws/gcp'
+
+  s.homepage = 'https://github.com/railsware/global'
+  s.licenses = ['MIT']
   s.summary = 'Simple way to load your configs from yaml/aws/gcp'
 
+  s.metadata['rubygems_mfa_required'] = 'true'
```
[the remark this answers is not in the document]
Comment (i7an) [reply global.gemspec:17]: @le0pard could you please share the script that does the release? I want to check that.

File: spec/global_merge_backends_spec.rb (hunk 12 above)
Comment (i7an) [reply spec/global_merge_backends_spec.rb:7]: in this particular case doubles are fine. global takes an object that quacks like a configuration. this spec tests merging, thus abstract doubles are legit.

File: global.gemspec (hunk 13 above)
[the remark this answers is not in the document]
Comment (le0pard) [reply global.gemspec:17]: Screenshot 2024-11-29 at 13 36 34

[the remark this answers is not in the document]
Comment (i7an) [reply global.gemspec:17]: I assumed there is CI that pushed a new version automatically. Ok.
Could you please check if MFA is on or off? If not maybe we should turn it on?

File: spec/global/backend/aws_parameter_store_spec.rb [hunk 14]
```diff
@@ -1,18 +1,18 @@
...
+
   let(:client) do
     Aws::SSM::Client.new(stub_responses: true)
   end
-  subject do
-    described_class.new(prefix: '/testapp/', client: client)
-  end
 
-  it 'reads parameters from the parameter store' do
+  it 'reads parameters from the parameter store' do # rubocop:disable RSpec/ExampleLength, RSpec/MultipleExpectations
```
[the remark this answers is not in the document]
Comment (i7an) [reply spec/global/backend/aws_parameter_store_spec.rb:15]: I tried several options. `next_token` expectations is the tricky part. If you want to keep them, you have to adjust rubocop configuration. To reduce the number of lines again you have to deal with `next_token` expectations, move stubs to elsewhere which does not improve code readability.
At the end I think disabling those two rules for one `it` is fine. If you disagree please suggest how to deal with these offences.

File: spec/global/backend/gcp_secret_manager_spec.rb [hunk 15]
```diff
@@ -1,39 +1,37 @@
...
-    @secret_data = double
-    allow(@secret_data).to receive_message_chain(:payload, :data).and_return('secret value')
+  before do
+    # rubocop:disable RSpec/VerifiedDoubles
+    match_item = double(name: 'prod-myapp-example-test_key')
+    not_match_item = double(name: 'different_key')
 
-    @not_match_item = double
-    allow(@not_match_item).to receive(:name).and_return('different_key')
+    secret_data = double(data: 'secret value')
```
Comment (i7an) [inline spec/global/backend/gcp_secret_manager_spec.rb:19]: 🤷‍♂️
```ruby
Google::Cloud::SecretManager::V1::SecretPayload.new.data
# => ""
instance_double(Google::Cloud::SecretManager::V1::SecretPayload, data: 'secret value')
# => RSpec::Mocks::MockExpectationError: class does not implement the instance method: data
```

**Review by le0pard — APPROVED** [review]

**Pull request merged**